include/shark/Algorithms/DirectSearch/CMSA.h

```diff
@@ -51,118 +51,131 @@
-		* \brief Calculates the center of gravity of the given population \f$ \in \mathbb{R}^d\f$.
-		*
-		* 
-		*/
-		template<typename Container, typename Extractor>
-		RealVector cog( const Container & container, const Extractor & e ) {
-
-			RealVector result( m_numberOfVariables, 0. );
-
-			for( std::size_t j = 0; j < container.size(); j++ )
-				result += 1./m_mu * e( container[j] );
-
-			return result;
-		}
-
-		SHARK_EXPORT_SYMBOL void read( InArchive & archive );
-		SHARK_EXPORT_SYMBOL void write( OutArchive & archive ) const;
-
-		using AbstractSingleObjectiveOptimizer<RealVector >::init;
-		/**
-		* \brief Initializes the algorithm for the supplied objective function.
-		*/
-		SHARK_EXPORT_SYMBOL void init( ObjectiveFunctionType& function, SearchPointType const& p);
-
-		/**
-		* \brief Executes one iteration of the algorithm.
-		*/
-		SHARK_EXPORT_SYMBOL void step(ObjectiveFunctionType const& function);
-
-		/**
-		* \brief Accesses the size of the parent population.
-		*/
-		std::size_t mu() const {
-			return m_mu;
-		}
-
-		/**
-		* \brief Accesses the size of the parent population, allows for l-value semantics.
-		*/
-		std::size_t & mu() {
-			return m_mu;
-		}
-
-		/**
-		* \brief Accesses the size of the offspring population.
-		*/
-		std::size_t lambda() const {
-			return m_lambda;
-		}
-
-		/**
-		* \brief Accesses the size of the offspring population, allows for l-value semantics.
-		*/
-		std::size_t & lambda() {
-			return m_lambda;
-		}
-	protected:
-		
-		std::size_t m_numberOfVariables; ///< Stores the dimensionality of the search space.
-		std::size_t m_mu; ///< The size of the parent population.
-		std::size_t m_lambda; ///< The size of the offspring population, needs to be larger than mu.
-
-		double m_sigma; ///< The current step size.
-		double m_cSigma; 
-		double m_cC; ///< Constant for adapting the covariance matrix.
-
-		RealVector m_mean; ///< The current cog of the population.
-
-		shark::MultiVariateNormalDistribution m_mutationDistribution; ///< Multi-variate normal mutation distribution.   
-	private:
-		/**
-		* \brief Updates the strategy parameters based on the supplied offspring population.
-		*/
-		SHARK_EXPORT_SYMBOL void updateStrategyParameters( const std::vector< IndividualType > & offspringNew ) ;
-	};
+	SHARK_EXPORT_SYMBOL void init( 
+		ObjectiveFunctionType& function, 
+		SearchPointType const& initialSearchPoint,
+		std::size_t lambda,
+		std::size_t mu,
+		double initialSigma,				       
+		const boost::optional< RealMatrix > & initialCovarianceMatrix = boost::optional< RealMatrix >()
+	);
+
+	/// \brief Executes one iteration of the algorithm.
+	SHARK_EXPORT_SYMBOL void step(ObjectiveFunctionType const& function);
+	
+	/// \brief sets the initial step length sigma
+	///
+	/// It is by default <=0 which means that sigma =1/sqrt(numVariables)
+	void setInitialSigma(double initSigma){
+		m_initSigma = initSigma;
+	}
+
+	/// \brief Sets the number of selected samples
+	void setMu(std::size_t mu){
```

Comment:
yes. it does make sense. the provided init method taking all the hyper parameters can not be provided as a virtual interface. so we need a way to set some parameters before init() is called(e.g. the user creates the optimizer, sets the hyper parameters and then provides it to the remaining program)